src/render/tutorials/graphql-usage/2_0.html.md

```diff
@@ -0,0 +1,777 @@
+    return new InputStreamReader(GraphQLHandlerFactory.class.getResourceAsStream("/" + path));
+}
+```
+
+Rememeber about updating services in META-INF. With a handler factory you can now configure a graphql endpoint. Add this path to `openapi.yml`:
+
+```yml
+/api/graphql:
+    post:
+        operationId: get-books-gql
+        responses:
+            default:
+                description: Books Graphql API
+```
+
+Note that it needs to be POST because GraphQL reads queries from the requests' bodies. Now add the operation to `operations.conf`:
+
+```HOCON
+{
+    operationId = get-books-gql
+    handlers = ${config.server.handlers.common.request} [
+        {
+            name = graphqlHandler
+            config = {
+                schema = "books.graphqls"
+            }
+        }
+    ]
+}
+```
+
+Note the `${config.server.handlers.common.request}` part. We import some standard configuration because we'll need it later. Now the only thing left are data fetchers and you'll have a working graphql endpoint.
+
+## Step 2
+We would like to fetch data from 3rd parties using existing Knot.x mechanisms, not to reinvent the wheel. Ideally we would have a generic data fetcher that can process Knot.x tasks defined in configuration. 
+
+Our fetcher will parse configuration to create tasks and run them using `FragmentsEngine`. It will output a `Fragment` that will contain data fetched from google in it's payload. Fragment's payload is just a `JsonObject` where actions can store data.
+
+
+#### Execute tasks from configuration
+Let's start with something like this:
+
+```java
+public abstract class TaskDataFetcher<T> implements DataFetcher<CompletableFuture<T>> {
+
+    private final Vertx vertx;
+    private final JsonObject config;
+    private final RoutingContext routingContext;
+    private final String taskName;
+    private final FragmentsEngine engine;
+
+    TaskDataFetcher(VertX, JsonObject config, RoutingContext routingContext, String taskName) {
+        this.vertx = vertx;
+        this.config = config;
+        this.routingContext = routingContext;
+        this.taskName = taskName;
+        engine = new FragmentsEngine(vertx);
+    }
+
+    @Override
+    public CompletableFuture<T> get(DataFetchingEnvironment environment) {
+        FragmentEventContextTaskAware eventContextTaskAware = setupTask(vertx, config, routingContext, environment);
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        engine
+            .execute(Collections.singletonList(eventContextTaskAware))
+            .subscribe(events -> {
+                // Here we'll get the outcome fragment's payload, parse it to model and complete the future with it
+            });
+        
+        return future;
+    }
+
+    //other methods we'll implement in a moment
+}
+```
+
+There's a couple of things to explain here. Firstly, our fetcher is abstract - it will have one abstract method for transforming payload into objects from our model (`Book`).

```

Comment:
```suggestion
There's a couple of things to explain here. First, our fetcher is abstract - it will have one abstract method for transforming payload into objects from our model (`Book`).
```

Reply:
I think `firstly` is fine. For reference: https://grammarist.com/usage/firstly-secondly-thirdly/

Reply:
Technically you're right. It just sounded weird to me. But I don't insist on changing that